src/modules/avatar/AvatarController.ts

```diff
@@ -0,0 +1,168 @@
+    private _inputMap : any;
+    private _idleAnim : Nullable<AnimationGroup> = null;
+    private _walkFwdAnim : Nullable<AnimationGroup> = null;
+    private _walkbwdAnim : Nullable<AnimationGroup> = null;
+    private _turnLeftAnim : Nullable<AnimationGroup> = null;
+    private _turnRightAnim : Nullable<AnimationGroup> = null;
+    private _currentAnim: Nullable<AnimationGroup> = null;
+    private _prevAnim: Nullable<AnimationGroup> = null;
+
+    constructor(avatar: Mesh, camera: ArcRotateCamera, scene: Scene, animGroups: AnimationGroup[]) {
+        this._avatar = avatar;
+        this._camera = camera;
+        this._scene = scene;
+        this._movement = new Vector3();
+        this._inputMap = {};
+
+        const nodes = new Map<string, Node>();
+        this._avatar.getChildren((node):boolean => {
+            nodes.set(node.name, node);
+            return true;
+        }, false);
+
+        animGroups.forEach((animGroup : AnimationGroup) => {
+            switch (animGroup.name) {
+                case "idle02":
+                    this._idleAnim = AvatarController._cloneAnimGroup(animGroup, nodes);
+                    break;
+                case "walk_fwd":
+                    this._walkFwdAnim = AvatarController._cloneAnimGroup(animGroup, nodes);
+                    break;
+                case "walk_bwd":
+                    this._walkbwdAnim = AvatarController._cloneAnimGroup(animGroup, nodes);
+                    break;
+                case "turn_left":
+                    this._turnLeftAnim = AvatarController._cloneAnimGroup(animGroup, nodes);
+                    break;
+                case "turn_right":
+                    this._turnRightAnim = AvatarController._cloneAnimGroup(animGroup, nodes);
+                    break;
+                default:
+            }
+
+        });
+
+    }
+
+    static _cloneAnimGroup(sourceAnimGroup : AnimationGroup, nodes : Map<string, Node>, loop = true):AnimationGroup {
+        const animGroup = new AnimationGroup(sourceAnimGroup.name);
+        animGroup.loopAnimation = loop;
+
+        sourceAnimGroup.targetedAnimations.forEach((targetAnim) => {
+            const target = targetAnim.target as TransformNode;
+            const node = nodes.get(target.name);
+            if (node) {
+                animGroup.addTargetedAnimation(targetAnim.animation, node);
+            }
+        });
+
+        return animGroup;
+    }
+
+    public start():void {
+        this._scene.registerBeforeRender(this._update.bind(this));
+
+        // scene action manager to detect inputs
+        if (!this._scene.actionManager) {
+            this._scene.actionManager = new ActionManager(this._scene);
+        }
+
+        this._scene.actionManager.registerAction(
+            new ExecuteCodeAction(ActionManager.OnKeyDownTrigger,
+                (evt) => {
+                    this._inputMap[evt.sourceEvent.key] = evt.sourceEvent.type === "keydown";
```

Comment:
Use `sourceEvent.code` instead of `sourceEvent.key`, here and elsewhere.

Reply:
Done.

Reply:
I also modify the key of show/hide to debugLayer shift + /